integration-tests/src/bin/read.rs

```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use libc::{read, write};
+
+fn main() {
+    let mut buf = [0u8; 16];
+    let mut sz: isize;
+    let mut i = 1;
+    loop {
+        let len = core::cmp::min(i, buf.len());
+        unsafe {
+            sz = read(libc::STDIN_FILENO, buf.as_mut_ptr() as *mut _, len);
+        }
+        if sz == 0 {
+            break;
+        }
+        unsafe {
+            write(libc::STDOUT_FILENO, buf.as_ptr() as *const _, sz as usize);
+        }
+        if sz < len as isize {
+            break;
+        }
+        i = 2 * i;
+    }
+}
```

Comment:
```suggestion
// SPDX-License-Identifier: Apache-2.0

use libc::{read, write, STDIN_FILENO as IN, STDOUT_FILENO as OUT};
use std::process::exit;
use std::cmp::min;

fn main() {
    let mut buf = [0u8; 16];
    let mut len = 1;

    loop {
        match unsafe { read(IN, buf.as_mut_ptr() as *mut _, len) } {
            sz if sz > 0 => unsafe { write(OUT, buf.as_ptr() as *const _, sz as usize); },
            sz if sz < 0 => exit(1),
            _ => break,
        }

        len = min(len * 2, buf.len());
    }
}
```

Take a few minutes to understand precisely why this works.

If `sz < len`, then we have EOF. But instead of immediately exiting, we run the loop one more time. On the next loop, `sz == 0` and we `break`.

Reply:
@npmccallum 
I took some time to understand how you used the match statement with place expressions. 
I get it now :)
Is there a reason why we are not libc::exit in case there is an error?

Reply:
`std::process::exit()` is the Rust way. But it probably just calls the `exit()` syscall under the covers.